### PR TITLE
[test] Setup python coverage test

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #1195

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

Current coverage = 67.03%
See https://codecov.io/gh/taichi-dev/taichi/src/master/python/taichi/lang/util.py, really cool!

### CODE REVIEW

**Comment by archibate** [comment]
C++ coverage setup may follow: https://stackoverflow.com/questions/44955539/how-to-properly-configure-codecov-for-a-c-w-cmake-project-in-travisci

**Comment by yuanming-hu** [comment]
This is awesome! Thanks for setting it up. Does codecov work for C++17? IIRC it only supports C++11/14.

**Comment by archibate** [comment]
I'm not sure, the search engine give me no result about codecov+c++11, not to talk c++17...
But I guess given that CI machines doesn't have GPU support, so online coverage report won't make sense anyway... Maybe we should keep C++ coverage offline like benchmark does? AFAIK `gcov` (a tool in gcc kit) can do this well, not sure it suits `clang`.

**Comment by k-ye** [comment]
Do we also need to update developer installation to mention this dependency? https://taichi.readthedocs.io/en/stable/dev_install.html#installing-depedencies

> Does codecov work for C++17? IIRC it only supports C++11/14.

Imagine setting this up and find out that C++ coverage is something like 0.2%...

**Comment by archibate** [comment]
> Do we also need to update developer installation to mention this dependency? https://taichi.readthedocs.io/en/stable/dev_install.html#installing-depedencies

Yes, btw, I found some of these deps not really required, like `scipy`, `psutils`, `opencv-python`.. I didn't install these on my machine but dev mode working correctly.

> > Does codecov work for C++17? IIRC it only supports C++11/14.
> 
> Imagine setting this up and find out that C++ coverage is something like 0.2%...

LOL, that's very possible given that Taichi have many legacy stuffs...

**Comment by yuanming-hu** [comment]
> Yes, btw, I found some of these deps not really required, like `scipy`, `psutils`, `opencv-python`.. I didn't install these on my machine but dev mode working correctly.

Yeah I think these can be removed... We used to need them but no longer now.

File: README.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 [![Build Status](https://img.shields.io/travis/taichi-dev/taichi?logo=Travis&branch=master)](https://travis-ci.com/taichi-dev/taichi)
 [![Build Status](https://img.shields.io/appveyor/build/yuanming-hu/taichi?logo=AppVeyor)](https://ci.appveyor.com/project/yuanming-hu/taichi/branch/master)
+[![Codecov Status](https://codecov.io/gh/taichi-dev/taichi/branch/master/graph/badge.svg)](https://codecov.io/gh/taichi-dev/taichi)
```
Comment (rexwangcc) [inline README.md]: One way to make it clear that the cpp coverage is not included in this report yet might be using shields.io over the default badge?

```suggestion
[![Codecov Status](https://img.shields.io/codecov/c/github/taichi-dev/taichi?label=Python%20Code%20Coverage&logo=codecov)](https://codecov.io/gh/taichi-dev/taichi)
```
which results in:

Codecov Status

File: docs/utilities.rst [hunk 2]
```diff
@@ -96,6 +96,26 @@ The suggested workflow for the performance related PR author to run the regressi
   then find the line of code triggering the error.
 
 
+.. _coverage:
+
+Code coverage
+-------------
+
+To ensure that our tests covered every situations, we need to **coverage report**.
```
Comment (rexwangcc) [inline docs/utilities.rst:104]:
```suggestion
To ensure that our tests covered every situations, we need to have **coverage report**.
```

File: misc/ci_setup.py [hunk 3]
```diff
@@ -185,11 +185,13 @@ def run(self):
...
             "yapf",
             "distro",
-            "pytest",
             "autograd",
             "astor",
+            "pytest",
             "pytest-xdist",
             "pytest-rerunfailures",
+            "pytest-cov",
+            "coverage",
```
[the remark this answers is not in the document]
Comment (archibate) [reply misc/ci_setup.py:194]: Yes, only `pytest` and `pytest-xdist` is required for end-user to `ti test`, they don't need to care about coverage anyway.

**Comment by archibate** [comment]
Thank for all your approvements! I'll merge once [CI](https://ci.appveyor.com/project/yuanming-hu/taichi/builds/33414445) passes, let's see if it's working well in master :)

**Pull request merged**